### PR TITLE
impl(bigtable): modern `AsyncRowReader`

### PR DESCRIPTION
**Pull request by dbolduc**
Most of the work for #9174

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9287)

### CODE REVIEW

**Review by coryan — COMMENTED** [review]
A few questions and nits.

It is really hard to get a sense of what the tests do, or how they changed.  The coverage looks good, but that is not a perfect way to assess if tests are good enough.  Any thoughts on that?

File: google/cloud/bigtable/internal/async_row_reader_test.cc [hunk 1]
```diff
@@ -12,1147 +12,916 @@
...
-
-  ASSERT_EQ(0U, cq_impl_->size());
+  AsyncRowReader::Create(cq, mock, kAppProfile, kTableName,
+                         on_row.AsStdFunction(), on_finish.AsStdFunction(),
+                         bigtable::RowSet(), bigtable::RowReader::NO_ROWS_LIMIT,
+                         bigtable::Filter::PassAllFilter(), std::move(retry),
+                         std::move(mock_b));
 }
 
-TEST_F(TableAsyncReadRowsTest, ReadRowNotFound) {
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/bigtable/internal/async_row_reader_test.cc:1102]: So this test fixture was originally for `Table::AsyncReadRow(s)`. (both singular and plural). Now, the test fixture is solely for unit testing the `bigtable_internal::AsyncRowReader` helper class.

These `ReadRow` tests are for [this logic](https://github.com/googleapis/google-cloud-cpp/blob/fd1d0ebe0e347c3074e8a90db94381863f3ccbbf/google/cloud/bigtable/table.cc#L435-L487), which will move to the `DataConnection`. We will test the logic in `data_connection_impl_test.cc`.

File: google/cloud/bigtable/internal/async_row_reader_test.cc [hunk 2]
```diff
@@ -12,1147 +12,916 @@
...
+              return make_ready_future(MakeResponse(std::move(v)));
+            })
+            .WillOnce([] { return make_ready_future(EndOfStream()); });
+        EXPECT_CALL(*stream, Finish).WillOnce([] {
+          return make_ready_future(Status{});
+        });
+        return stream;
+      });
 
-TEST_F(TableAsyncReadRowsTest, ReadRowError) {
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/bigtable/internal/async_row_reader_test.cc:1128]: Same answer as: https://github.com/googleapis/google-cloud-cpp/pull/9287#discussion_r899718671

**Review by coryan — APPROVED** [review]

File: google/cloud/bigtable/internal/async_row_reader_test.cc [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2019 Google LLC
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/bigtable/internal/async_row_reader_test.cc]: Fixed. Changing the license year was unintentional. Thanks for catching it.

File: google/cloud/bigtable/internal/async_row_reader_test.cc [hunk 4]
```diff
@@ -12,1147 +12,916 @@
...
-      })
-      .RetiresOnSaturation();
-  EXPECT_CALL(stream, Finish).WillOnce([](grpc::Status* status, void*) {
-    *status = grpc::Status::OK;
-  });
-
-  ExpectRow("r1");
-  promises_from_user_cb_[0].set_value(true);
-  ReadRows(1);
+TEST(AsyncRowReaderTest, ParserEofDoesntFailsOnUnfinishedRowIfRowLimit) {
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/bigtable/internal/async_row_reader_test.cc]: Done.

File: google/cloud/bigtable/internal/async_row_reader_test.cc [hunk 5]
```diff
@@ -12,1147 +12,916 @@
...
 
 /// @test Verify that permanent errors are not retried and properly passed.
-TEST_F(TableAsyncReadRowsTest, PermanentFailure) {
-  auto& stream = AddReader([](btproto::ReadRowsRequest const&) {});
+TEST(AsyncRowReaderTest, PermanentFailure) {
+  CompletionQueue cq;
+
+  auto mock = std::make_shared<MockBigtableStub>();
+  EXPECT_CALL(*mock, AsyncReadRows)
+      .WillOnce([](CompletionQueue const&, std::unique_ptr<grpc::ClientContext>,
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/bigtable/internal/async_row_reader_test.cc]: Neat.

**Comment by dbolduc** [comment]
> It is really hard to get a sense of what the tests do

All the tests have the same structure, so I added comments for the first one. I also wrote in sentences the expectations we are looking for, for the tests that are less obvious. I hope that is what you mean.

>or how they changed.

They hardly look alike. I think mainly due to using a MockStream vs. running everything on a FakeCompletionQueueImpl.

> The coverage looks good, but that is not a perfect way to assess if tests are good enough. Any thoughts on that?

I looked through the code again. I added two tests for code that we were not exercising. I think what we have is good enough.

**Pull request merged**